flo-runner/src/main/java/com/spotify/flo/context/ForkingEvalContext.java

```diff
@@ -0,0 +1,382 @@
+/*-
+ * -\-\-
+ * Flo Runner
+ * --
+ * Copyright (C) 2016 - 2018 Spotify AB
+ * --
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * 
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * -/-/-
+ */
+
+package com.spotify.flo.context;
+
+import com.spotify.flo.EvalContext;
+import com.spotify.flo.Fn;
+import com.spotify.flo.TaskId;
+import com.spotify.flo.Tracing;
+import com.spotify.flo.freezer.PersistingContext;
+import io.grpc.Context;
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.PrintStream;
+import java.lang.management.ManagementFactory;
+import java.nio.file.FileVisitResult;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.SimpleFileVisitor;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.time.LocalTime;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * An {@link EvalContext} that runs tasks in sub-processes.
+ * <p>
+ * Forking can be disabled using the environment variable {@code FLO_DISABLE_FORKING=true}.
+ * <p>
+ * Forking is disabled by default when running in the debugger, but can be enabled by {@code FLO_FORCE_FORK=true}.
+ */
+class ForkingEvalContext extends ForwardingEvalContext {
+
+  private static final Logger log = LoggerFactory.getLogger(ForwardingEvalContext.class);
+
+  // Is the Java Debug Wire Protocol activated?
+  private static boolean IN_DEBUGGER = ManagementFactory.getRuntimeMXBean().
+      getInputArguments().stream().anyMatch(s -> s.contains("-agentlib:jdwp"));
+
+  private static boolean FLO_DISABLE_FORKING = Boolean.parseBoolean(System.getenv("FLO_DISABLE_FORKING"));
+
+  private static boolean FORCE_FORK = Boolean.parseBoolean(System.getenv("FLO_FORCE_FORK"));
```

Comment:
Wouldn't it be cleaner to just deal with this in `FloRunner` and skip adding `ForkingEvalContext` in the chain? That would more clearly separate concerns.

Reply:
How can we avoid adding `ForkingEvalContext` in the chain? I'm not sure I understand this.

Reply:
In https://github.com/spotify/flo/blob/master/flo-runner/src/main/java/com/spotify/flo/context/FloRunner.java#L153, there's some logic already that skips adding the PersistingContext, or selecting the base sync/async context. Using the ForkingEvalContext could be added (or not) in the same way).

Reply:
I'll take a stab at it, perhaps by putting the check in `composeWith` or a `maybeComposeWith`. Or would it make more sense to put the environment variable lookups and decisions etc in FloRunner?